File: invar/detmath.py
# Copyright (c) 2026 Anomly, Inc. All rights reserved. Author: Ry Bruscoe.
"""detmath: the ggml-det deterministic elementwise library, in Python, bit-exact.

ggml-det (llama-cpp-et, ggml/src/ggml-det/ggml-det.h) defines every non-matmul op of the
exact profile as a fixed sequence of IEEE round-to-nearest double / float32 operations
with no libm and no fused multiply-add. Python floats are IEEE doubles with the same
rounding and no contraction, so the same sequence gives the same bits; float32 steps are
rounded through struct. Conformance is checked against the C library (tests/test_detmath.py).
"""
from __future__ import annotations

import math
import struct
from fractions import Fraction

_INF = float("inf")


_F32_MAX = 3.4028234663852886e+38
_F32_OVF = 3.4028235677973366e+38     # values >= this round to +inf (halfway to the next binade)


def f32(x: float) -> float:
    """Round a double to float32 (round-to-nearest-even), returned as a Python float;
    magnitudes beyond the float32 range become +-inf exactly as a C (float) cast does."""
    if x != x or x in (_INF, -_INF):
        return x
    if x >= _F32_OVF:
        return _INF
    if x <= -_F32_OVF:
        return -_INF
    if x > _F32_MAX:
        return _F32_MAX
    if x < -_F32_MAX:
        return -_F32_MAX
    return struct.unpack("<f", struct.pack("<f", x))[0]


def f32_bits(x: float) -> int:
    return struct.unpack("<I", struct.pack("<f", x))[0]


def f32_from_bits(b: int) -> float:
    return struct.unpack("<f", struct.pack("<I", b))[0]


def fmul(a: float, b: float) -> float:
    """float32 multiply: a, b are float32 values; the double product rounded to float32 is
    the correctly rounded float32 product (double rounding is innocuous for *, +, -, /, sqrt)."""
    return f32(a * b)


def fadd(a: float, b: float) -> float:
    return f32(a + b)


def fsub(a: float, b: float) -> float:
    return f32(a - b)


def fdiv(a: float, b: float) -> float:
    return f32(a / b)


def fsqrt(a: float) -> float:
    return f32(math.sqrt(a))


# ----------------------------------------------------------------------------- helpers

def det_ldexp(x: float, k: int) -> float:
    return math.ldexp(x, k)          # exact (single rounding on subnormal results), as in C


def det_floor(x: float) -> float:
    t = int(x)                        # truncation toward zero
    r = float(t)
    if r > x:
        r = float(t - 1)
    return r


# ----------------------------------------------------------------------------- exact sums

def sumsq_f32_exact(xs) -> float:
    """Exact sum of squares of float32 values, correctly rounded to double (det_sumsq_f32);
    any non-finite element -> +inf."""
    S = Fraction(0)
    for v in xs:
        if v == 0.0:
            continue
        if not math.isfinite(v):
            return _INF
        f = Fraction(v)
        S += f * f
    return float(S)                    # Fraction -> float is correctly rounded


def rms_scale(sumsq: float, n: int, eps: float) -> float:
    """det_rms_scale: mean = (float)(S/n) in double; scale = 1/sqrtf(mean + eps)."""
    mean = f32(sumsq / float(n))
    return fdiv(1.0, fsqrt(fadd(mean, eps)))


def rms_norm_row(x, w, eps: float):
    """RMSNorm followed by the weight multiply, as both backends compute it: y = (x*scale)*w."""
    scale = rms_scale(sumsq_f32_exact(x), len(x), eps)
    return [fmul(fmul(xi, scale), wi) for xi, wi in zip(x, w)]


# ----------------------------------------------------------------------------- exp / log / trig

_INV_LN2 = 1.44269504088896338700e+00
_LN2_HI = 6.93147180369123816490e-01
_LN2_LO = 1.90821492927058770002e-10
_LN2 = 6.93147180559945286227e-01
_EXP_C = [1.0, 1.0, 5.00000000000000000000e-01, 1.66666666666666657415e-01, 4.16666666666666643537e-02,
          8.33333333333333321769e-03, 1.38888888888888894189e-03, 1.98412698412698412526e-04,
          2.48015873015873015658e-05, 2.75573192239858925110e-06, 2.75573192239858883130e-07,
          2.50521083854417202234e-08, 2.08767569878681002718e-09, 1.60590438368216133364e-10]


def exp_small(r: float) -> float:
    p = _EXP_C[13]
    for i in range(12, -1, -1):
        p = p * r + _EXP_C[i]
    return p


def exp_d(x: float) -> float:
    k = det_floor(x * _INV_LN2 + 0.5)
    r = (x - k * _LN2_HI) - k * _LN2_LO
    return det_ldexp(exp_small(r), int(k))


def expf(x: float) -> float:
    """det_expf on a float32 value."""
    b = f32_bits(x)
    if ((b >> 23) & 0xFF) == 0xFF:
        if b & 0x7FFFFF:
            return x                   # nan
        return 0.0 if (b >> 31) else x  # -inf -> 0, +inf -> inf
    if x > f32(88.75):
        return _INF
    if x < -104.0:
        return 0.0
    return f32(exp_d(x))


def exp2_d(y: float) -> float:
    k = det_floor(y + 0.5)
    r = y - k
    return det_ldexp(exp_small(r * _LN2), int(k))


def log2_d(x: float) -> float:
    m, e = math.frexp(x)               # x = m * 2^e, m in [0.5, 1)
    m *= 2.0                           # [1, 2), exact
    e -= 1
    if m > 1.41421356237309514547e+00:
        m = m * 0.5
        e += 1
    s = (m - 1.0) / (m + 1.0)
    s2 = s * s
    p = 0.0
    for k in range(29, 0, -2):
        p = p * s2 + 1.0 / float(k)
    ln_m = 2.0 * (s * p)
    return float(e) + ln_m * _INV_LN2


_TWO_OVER_PI = 6.36619772367581382433e-01
_PIO2_1 = 1.57079632673412561417e+00
_PIO2_2 = 6.07710050650619224932e-11
_PIO2_3 = 2.02226624879595063154e-21
_SC = [1.0, -1.66666666666666657415e-01, 8.33333333333333321769e-03, -1.98412698412698412526e-04,
       2.75573192239858925110e-06, -2.50521083854417202234e-08, 1.60590438368216133364e-10,
       -7.64716373181981647590e-13]
_CC = [1.0, -5.00000000000000000000e-01, 4.16666666666666643537e-02, -1.38888888888888894189e-03,
       2.48015873015873015658e-05, -2.75573192239858883130e-07, 2.08767569878681002718e-09,
       -1.14707455977297245139e-11, 4.77947733238738529744e-14]


def sincos_d(x: float) -> tuple[float, float]:
    n = det_floor(x * _TWO_OVER_PI + 0.5)
    r = ((x - n * _PIO2_1) - n * _PIO2_2) - n * _PIO2_3
    r2 = r * r
    ps = _SC[7]
    for i in range(6, -1, -1):
        ps = ps * r2 + _SC[i]
    ps = ps * r
    pc = _CC[8]
    for i in range(7, -1, -1):
        pc = pc * r2 + _CC[i]
    q = int(n) & 3
    if q == 0:
        return ps, pc
    if q == 1:
        return pc, -ps
    if q == 2:
        return -ps, -pc
    return -pc, ps


def sincosf(theta: float) -> tuple[float, float]:
    s, c = sincos_d(theta)
    return f32(s), f32(c)


def rope_freq(i: int, n_dims: int, freq_base: float) -> float:
    L = log2_d(float(freq_base)) * ((-2.0 * float(i)) / float(n_dims))
    return exp2_d(L)


def rope_sincos(pos: float, i: int, n_dims: int, freq_base: float, freq_scale: float) -> tuple[float, float]:
    theta = (float(pos) * rope_freq(i, n_dims, freq_base)) * float(freq_scale)
    s, c = sincos_d(theta)
    return f32(s), f32(c)


def rope_sincos_ff(pos: float, i: int, n_dims: int, freq_base: float, freq_scale: float, ff: float) -> tuple[float, float]:
    """RoPE with a per-pair frequency factor (Llama 3 rope_freqs): theta = ((pos*freq)/ff)*freq_scale."""
    theta = ((float(pos) * rope_freq(i, n_dims, freq_base)) / float(ff)) * float(freq_scale)
    s, c = sincos_d(theta)
    return f32(s), f32(c)


def tanh_d(y: float) -> float:
    if y > 20.0:
        return 1.0
    if y < -20.0:
        return -1.0
    u = exp_d(2.0 * y)
    return (u - 1.0) / (u + 1.0)


def tanhf(x: float) -> float:
    if x != x:
        return x
    return f32(tanh_d(x))


_GELU_COEF_A = f32(0.044715)
_SQRT_2_OVER_PI = f32(0.79788456080286535587989211986876)


def geluf(x: float) -> float:
    """ggml's tanh-GELU with its operation order: 0.5f*x*(1 + tanhf(S*x*(1 + A*x*x)))."""
    inner = fmul(_SQRT_2_OVER_PI, fmul(x, fadd(1.0, fmul(_GELU_COEF_A, fmul(x, x)))))
    return fmul(fmul(0.5, x), fadd(1.0, tanhf(inner)))


def sigmoidf(x: float) -> float:
    return fdiv(1.0, fadd(1.0, expf(-x)))


def siluf(x: float) -> float:
    return fmul(x, sigmoidf(x))


# ----------------------------------------------------------------------------- composite ops

def rope_row(x, pos: int, n_dims: int, freq_base: float, freq_scale: float, attn_factor: float = 1.0,
             neox: bool = False, freq_factors=None):
    """RoPE of one row (float32 values) as both backends compute it: cos/sin from
    rope_sincos (with the pair's frequency factor when the model has rope_freqs) times
    attn_factor, rotation x0*c - x1*s, x0*s + x1*c in float32."""
    y = list(x)
    half = n_dims // 2
    for i in range(half):
        if freq_factors is not None:
            s, c = rope_sincos_ff(pos, i, n_dims, freq_base, freq_scale, freq_factors[i])
        else:
            s, c = rope_sincos(pos, i, n_dims, freq_base, freq_scale)
        c = fmul(c, attn_factor)
        s = fmul(s, attn_factor)
        if neox:
            i0, i1 = i, i + half
        else:
            i0, i1 = 2 * i, 2 * i + 1
        x0, x1 = x[i0], x[i1]
        y[i0] = fsub(fmul(x0, c), fmul(x1, s))
        y[i1] = fadd(fmul(x0, s), fmul(x1, c))
    return y


def swiglu_row(gate, up):
    return [fmul(siluf(g), u) for g, u in zip(gate, up)]


def geglu_row(gate, up):
    return [fmul(geluf(g), u) for g, u in zip(gate, up)]


def add_row(a, b):
    return [fadd(x, y) for x, y in zip(a, b)]


def soft_max_row(x, scale: float = 1.0, mask=None):
    """Softmax as both backends compute it: v = x*scale + mask; e = expf(v - max);
    p = e * (1 / (float) exact_sum(e))."""
    v = [fmul(xi, scale) for xi in x]
    if mask is not None:
        v = [fadd(vi, mi) for vi, mi in zip(v, mask)]
    mx = max(v)
    e = [expf(fsub(vi, mx)) for vi in v]
    S = Fraction(0)
    for ei in e:
        S += Fraction(ei)
    inv = fdiv(1.0, f32(float(S)))
    return [fmul(ei, inv) for ei in e]

File: go/crverify/detmath.go
// Copyright (c) 2026 Anomly, Inc. All rights reserved. Author: Ry Bruscoe.
// Licensed under the Apache License, Version 2.0 (same as the invar repository).

package crverify

// detmath: the ggml-det deterministic elementwise library in Go, bit-exact with the C header
// and the Python port (invar/detmath.py). Go float64 arithmetic is IEEE round-to-nearest; the
// compiler may fuse x*y+z on some architectures (arm64), so every product is passed through an
// explicit float64() conversion, which the Go specification defines as a rounding barrier.

import "math"

func dmul(a, b float64) float64 { return float64(a * b) }
func dadd(a, b float64) float64 { return float64(a + b) }
func dsub(a, b float64) float64 { return float64(a - b) }
func ddiv(a, b float64) float64 { return float64(a / b) }

// float32 ops: the double result rounded to float32 is the correctly rounded float32 result
// (double rounding is innocuous for +, -, *, / and sqrt at these precisions).
func fmul(a, b float32) float32 { return float32(float64(a) * float64(b)) }
func fadd(a, b float32) float32 { return float32(float64(a) + float64(b)) }
func fsub(a, b float32) float32 { return float32(float64(a) - float64(b)) }
func fdiv(a, b float32) float32 { return float32(float64(a) / float64(b)) }
func fsqrt(a float32) float32   { return float32(math.Sqrt(float64(a))) }

func detFloor(x float64) float64 {
	t := int64(x)
	r := float64(t)
	if r > x {
		r = float64(t - 1)
	}
	return r
}

// ---------------------------------------------------------------- exact sums (640-bit, radix 352)

const detBigLimbs = 20
const detBigRadix = 352

type detBig [detBigLimbs]uint32

func (a *detBig) addPiece(w int, parts [3]uint32) {
	var c uint64
	for i := 0; i < 3; i++ {
		if w+i >= detBigLimbs {
			return
		}
		t := uint64(a[w+i]) + uint64(parts[i]) + c
		a[w+i] = uint32(t)
		c = t >> 32
	}
	for i := w + 3; c != 0 && i < detBigLimbs; i++ {
		t := uint64(a[i]) + c
		a[i] = uint32(t)
		c = t >> 32
	}
}

func (a *detBig) addScaled(mi uint64, pos int) {
	w, b := pos>>5, uint(pos&31)
	lo, hi := mi, uint64(0)
	if b != 0 {
		lo = mi << b
		hi = mi >> (64 - b)
	}
	a.addPiece(w, [3]uint32{uint32(lo), uint32(lo >> 32), uint32(hi)})
}

// AddSq: acc += v*v exactly; false for a non-finite v
func (a *detBig) AddSq(v float32) bool {
	vd := float64(v)
	if vd == 0 {
		return true
	}
	fb := math.Float32bits(v)
	if (fb>>23)&0xFF == 0xFF {
		return false
	}
	p := dmul(vd, vd)
	bits := math.Float64bits(p)
	mi := (bits & 0xFFFFFFFFFFFFF) | (1 << 52)
	e2 := int((bits>>52)&0x7FF) - 1023
	a.addScaled(mi, e2-52+detBigRadix)
	return true
}

// AddNonNeg: acc += v for a finite non-negative float32
func (a *detBig) AddNonNeg(v float32) {
	fb := math.Float32bits(v)
	e, f := int((fb>>23)&0xFF), uint64(fb&0x7FFFFF)
	if fb&0x7FFFFFFF == 0 || e == 0xFF {
		return
	}
	var mi uint64
	var e2 int
	if e == 0 {
		mi, e2 = f, -126-23
	} else {
		mi, e2 = f|0x800000, e-127-23
	}
	a.addScaled(mi, e2+detBigRadix)
}

func (a *detBig) top() int {
	for i := detBigLimbs - 1; i >= 0; i-- {
		if a[i] != 0 {
			t := 31
			for (a[i]>>uint(t))&1 == 0 {
				t--
			}
			return 32*i + t
		}
	}
	return -1
}

func (a *detBig) bit(j int) uint32 {
	if j < 0 || j >= 32*detBigLimbs {
		return 0
	}
	return (a[j>>5] >> uint(j&31)) & 1
}

// ToDouble: correctly rounded (nearest-even) value S * 2^-352
func (a *detBig) ToDouble() float64 {
	t := a.top()
	if t < 0 {
		return 0
	}
	var mant uint64
	for j := t; j > t-53; j-- {
		mant = mant<<1 | uint64(a.bit(j))
	}
	guard := a.bit(t - 53)
	sticky := uint32(0)
	for j := t - 54; j >= 0 && sticky == 0; j-- {
		sticky |= a.bit(j)
	}
	e := t - 52
	if guard == 1 && (sticky == 1 || mant&1 == 1) {
		mant++
		if mant == 1<<53 {
			mant >>= 1
			e++
		}
	}
	return math.Ldexp(float64(mant), e-detBigRadix)
}

// SumsqF32 is det_sumsq_f32; RmsScale is det_rms_scale.
func SumsqF32(x []float32) float64 {
	var acc detBig
	for _, v := range x {
		if !acc.AddSq(v) {
			return math.Inf(1)
		}
	}
	return acc.ToDouble()
}

func RmsScale(sumsq float64, n int, eps float32) float32 {
	mean := float32(ddiv(sumsq, float64(n)))
	return fdiv(1, fsqrt(fadd(mean, eps)))
}

func RmsNormRow(x, w []float32, eps float32) []float32 {
	scale := RmsScale(SumsqF32(x), len(x), eps)
	out := make([]float32, len(x))
	for i := range x {
		out[i] = fmul(fmul(x[i], scale), w[i])
	}
	return out
}

// ---------------------------------------------------------------- exp / log2 / exp2 / trig

const (
	detInvLn2 = 1.44269504088896338700e+00
	detLn2Hi  = 6.93147180369123816490e-01
	detLn2Lo  = 1.90821492927058770002e-10
	detLn2    = 6.93147180559945286227e-01
)

var detExpC = [14]float64{1.0, 1.0, 5.00000000000000000000e-01, 1.66666666666666657415e-01, 4.16666666666666643537e-02,
	8.33333333333333321769e-03, 1.38888888888888894189e-03, 1.98412698412698412526e-04,
	2.48015873015873015658e-05, 2.75573192239858925110e-06, 2.75573192239858883130e-07,
	2.50521083854417202234e-08, 2.08767569878681002718e-09, 1.60590438368216133364e-10}

func detExpSmall(r float64) float64 {
	p := detExpC[13]
	for i := 12; i >= 0; i-- {
		p = dadd(dmul(p, r), detExpC[i])
	}
	return p
}

func DetExpD(x float64) float64 {
	k := detFloor(dadd(dmul(x, detInvLn2), 0.5))
	r := dsub(dsub(x, dmul(k, detLn2Hi)), dmul(k, detLn2Lo))
	return math.Ldexp(detExpSmall(r), int(k))
}

func DetExpf(x float32) float32 {
	b := math.Float32bits(x)
	if (b>>23)&0xFF == 0xFF {
		if b&0x7FFFFF != 0 {
			return x
		}
		if b>>31 == 1 {
			return 0
		}
		return x
	}
	if x > 88.75 {
		return float32(math.Inf(1))
	}
	if x < -104.0 {
		return 0
	}
	return float32(DetExpD(float64(x)))
}

func DetExp2D(y float64) float64 {
	k := detFloor(dadd(y, 0.5))
	r := dsub(y, k)
	return math.Ldexp(detExpSmall(dmul(r, detLn2)), int(k))
}

func DetLog2D(x float64) float64 {
	bits := math.Float64bits(x)
	e := int((bits>>52)&0x7FF) - 1023
	m := math.Float64frombits((bits & 0xFFFFFFFFFFFFF) | (1023 << 52))
	if m > 1.41421356237309514547e+00 {
		m = dmul(m, 0.5)
		e++
	}
	s := ddiv(dsub(m, 1.0), dadd(m, 1.0))
	s2 := dmul(s, s)
	p := 0.0
	for k := 29; k >= 1; k -= 2 {
		p = dadd(dmul(p, s2), ddiv(1.0, float64(k)))
	}
	lnm := dmul(2.0, dmul(s, p))
	return dadd(float64(e), dmul(lnm, detInvLn2))
}

const (
	detTwoOverPi = 6.36619772367581382433e-01
	detPio2_1    = 1.57079632673412561417e+00
	detPio2_2    = 6.07710050650619224932e-11
	detPio2_3    = 2.02226624879595063154e-21
)

var detSC = [8]float64{1.0, -1.66666666666666657415e-01, 8.33333333333333321769e-03, -1.98412698412698412526e-04,
	2.75573192239858925110e-06, -2.50521083854417202234e-08, 1.60590438368216133364e-10, -7.64716373181981647590e-13}
var detCC = [9]float64{1.0, -5.00000000000000000000e-01, 4.16666666666666643537e-02, -1.38888888888888894189e-03,
	2.48015873015873015658e-05, -2.75573192239858883130e-07, 2.08767569878681002718e-09, -1.14707455977297245139e-11,
	4.77947733238738529744e-14}

func DetSincosD(x float64) (s, c float64) {
	n := detFloor(dadd(dmul(x, detTwoOverPi), 0.5))
	r := dsub(dsub(dsub(x, dmul(n, detPio2_1)), dmul(n, detPio2_2)), dmul(n, detPio2_3))
	r2 := dmul(r, r)
	ps := detSC[7]
	for i := 6; i >= 0; i-- {
		ps = dadd(dmul(ps, r2), detSC[i])
	}
	ps = dmul(ps, r)
	pc := detCC[8]
	for i := 7; i >= 0; i-- {
		pc = dadd(dmul(pc, r2), detCC[i])
	}
	switch int64(n) & 3 {
	case 0:
		return ps, pc
	case 1:
		return pc, -ps
	case 2:
		return -ps, -pc
	default:
		return -pc, ps
	}
}

func RopeFreq(i, nDims int, freqBase float32) float64 {
	L := dmul(DetLog2D(float64(freqBase)), ddiv(dmul(-2.0, float64(i)), float64(nDims)))
	return DetExp2D(L)
}

func RopeSincos(pos float32, i, nDims int, freqBase, freqScale float32) (s, c float32) {
	theta := dmul(dmul(float64(pos), RopeFreq(i, nDims, freqBase)), float64(freqScale))
	sd, cd := DetSincosD(theta)
	return float32(sd), float32(cd)
}

// RopeSincosFF: RoPE with a per-pair frequency factor (Llama 3 rope_freqs).
func RopeSincosFF(pos float32, i, nDims int, freqBase, freqScale, ff float32) (s, c float32) {
	theta := dmul(ddiv(dmul(float64(pos), RopeFreq(i, nDims, freqBase)), float64(ff)), float64(freqScale))
	sd, cd := DetSincosD(theta)
	return float32(sd), float32(cd)
}

func Sigmoidf(x float32) float32 { return fdiv(1, fadd(1, DetExpf(-x))) }
func Siluf(x float32) float32    { return fmul(x, Sigmoidf(x)) }

// RopeRow rotates one head-dim slice in place semantics (returns a new slice).
func RopeRow(x []float32, pos int, nDims int, freqBase, freqScale, attnFactor float32, neox bool, ff []float32) []float32 {
	y := make([]float32, len(x))
	copy(y, x)
	half := nDims / 2
	for i := 0; i < half; i++ {
		var s, c float32
		if ff != nil {
			s, c = RopeSincosFF(float32(pos), i, nDims, freqBase, freqScale, ff[i])
		} else {
			s, c = RopeSincos(float32(pos), i, nDims, freqBase, freqScale)
		}
		c = fmul(c, attnFactor)
		s = fmul(s, attnFactor)
		var i0, i1 int
		if neox {
			i0, i1 = i, i+half
		} else {
			i0, i1 = 2*i, 2*i+1
		}
		x0, x1 := x[i0], x[i1]
		y[i0] = fsub(fmul(x0, c), fmul(x1, s))
		y[i1] = fadd(fmul(x0, s), fmul(x1, c))
	}
	return y
}

func SwigluRow(gate, up []float32) []float32 {
	out := make([]float32, len(gate))
	for i := range gate {
		out[i] = fmul(Siluf(gate[i]), up[i])
	}
	return out
}

func AddRow(a, b []float32) []float32 {
	out := make([]float32, len(a))
	for i := range a {
		out[i] = fadd(a[i], b[i])
	}
	return out
}

// SoftMaxRow: v = x*scale + mask; e = expf(v - max); p = e * (1 / (float32) exact_sum(e))
func SoftMaxRow(x []float32, scale float32, mask []float32) []float32 {
	v := make([]float32, len(x))
	mx := float32(math.Inf(-1))
	for i := range x {
		v[i] = fmul(x[i], scale)
		if mask != nil {
			v[i] = fadd(v[i], mask[i])
		}
		if v[i] > mx {
			mx = v[i]
		}
	}
	var acc detBig
	e := make([]float32, len(x))
	for i := range v {
		e[i] = DetExpf(fsub(v[i], mx))
		acc.AddNonNeg(e[i])
	}
	inv := fdiv(1, float32(acc.ToDouble()))
	for i := range e {
		e[i] = fmul(e[i], inv)
	}
	return e
}
